sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/cache/CacheFactory.java

```diff
@@ -0,0 +1,49 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.io.gcp.spanner.changestreams.cache;
+
+import java.io.Serializable;
+import javax.annotation.Nullable;
+import org.apache.beam.sdk.io.gcp.spanner.changestreams.dao.DaoFactory;
+import org.joda.time.Duration;
+
+@SuppressWarnings({"initialization.fields.uninitialized", "nullness"})
+public class CacheFactory implements Serializable {
+
+  private static final long serialVersionUID = -8722905670370252723L;
+  private final DaoFactory daoFactory;
+  private final @Nullable Duration refreshRate;
+  private transient WatermarkCache watermarkCacheInstance;
+
+  public CacheFactory(DaoFactory daoFactory, @Nullable Duration watermarkRefreshRate) {
+    this.daoFactory = daoFactory;
+    this.refreshRate = watermarkRefreshRate;
+  }
+
+  public synchronized WatermarkCache getWatermarkCache() {
+    if (watermarkCacheInstance == null) {
+      if (refreshRate == null) {
+        watermarkCacheInstance = new NoOpWatermarkCache(daoFactory.getPartitionMetadataDao());
```

Comment:
alternatively you could set this in the constructor and in readObject

https://stackoverflow.com/questions/4119181/java-object-serialization-readobject-defaultreadobject

Reply:
Obsolete now that we have a static concurrent hashmap